corehq/util/test_utils.py

```diff
@@ -169,6 +170,40 @@ class flag_disabled(flag_enabled):
     enabled = False
 
 
+@contextmanager
+def disable_overpatch(patch_obj):
+    """Prevent class decorator from masking method decorator"""
+    def apply_to(patch):
+        def decorate_callable(func):
+            key = getkey(patch)
+            if any(key == getkey(p) for p in getattr(func, "patchings", [])):
+                # do not decorate if already patched for this oneshot
+                return func
+            return real_decorate_callable(func)
+
+        def copy():
+            return apply_to(real_copy())
+
+        real_decorate_callable = patch.decorate_callable
+        real_copy = patch.copy
+        patch.decorate_callable = decorate_callable
+        patch.copy = copy
+        patches.append(patch)
+        return patch
+
+    def getkey(patch):
+        return patch.getter(), patch.attribute
+
+    patches = []
+    apply_to(patch_obj)
+    try:
+        yield
+    finally:
+        for patch in patches:
+            del patch.decorate_callable
+            del patch.copy
```

Comment:
Why delete these rather than set them back to the original value?

Reply:
`apply_to` sets attributes on the `patch` instance, which masks class methods with the same name. Deleting the attributes restores the previous state of having the methods defined on the class.

Reply:
Ah like this?
```python
In [1]: class Foo:
   ...:     @classmethod
   ...:     def bar(cls):
   ...:         return 'original'
   ...: 

In [2]: f = Foo()

In [3]: f.bar = lambda: 'overridden'

In [4]: f.bar()
Out[4]: 'overridden'

In [5]: del f.bar

In [6]: f.bar()
Out[6]: 'original'
```
Makes sense, thanks!